### PR TITLE
fix: in case txhashset validation fail, 'TxHashsetDownload' should retry (#1293)

### PR DESCRIPTION
**Pull request by garyyu**
1. To communicate the sync errors across parts of the system, enhance the `SyncState` structure to have a `sync_error` element.
2. In `txhashset_write()`, set this `sync_error` in case of failure. 
3. In `run_sync` thread, read this `sync_error` and give proper handling: for example in this case restart TxHashset download.

And one more related fix:

- When `header_sync_due` is due, must always launch `header_sync()`, otherwise, `if header_head.height == si.highest_height` could always fail and cause `fast_sync_due()` never executed.

### CODE REVIEW

**Comment by hashmap** [comment]
Thanks for the PR! I'm not sure about the design, `sync_error` is like a second dimension for the state machine we have, what advantages of it do you see? Have you considered adding a state  to `SyncStatus` and handling it here for example https://github.com/mimblewimble/grin/blob/master/chain/src/chain.rs#L553  after extending the trait https://github.com/mimblewimble/grin/blob/master/chain/src/types.rs#L125 ?
Also a txhashet request may be ignored by a peer, or just fail, we need to handle this case too.

**Comment by garyyu** [comment]
@hashmap  Thanks for the review~  below are answers to your questions.

Question 1. `sync_error is like a second dimension for the state machine we have,  what advantages of it do you see?`
Answer: 
- The advantage is to communicate `errors` only, not directly touch the `SyncState` change, that's a way to avoid coupling between different modules. 
- Another advantage is you can extend all kinds of error handling with this `sync_error`, not only for this fix, also for any future error handling.

Question 2. `Have you considered adding a state to SyncStatus and handling it ?`
Answer: 
- compare to add a `sync_error`, add one or more `SyncStatus` will need change that status directly, other module perhaps is using this status, that's a deep coupling.
- in future, if other errors need to be handled, you have to add more and more `SyncStatus`, but actually, the errors are already there, we can just reuse the errors.

Question 3. `Also a txhashet request may be ignored by a peer, or just fail, we need to handle this case too.`
Answer: very good question! Do you mean the `MessageHandler for Protocol` in protocol.rs ? What's the indicator of a request ignored or fail?

**Comment by hashmap** [comment]
@garyyu thanks! I'd prefer to rely on state changes only, but it's a matter of taste, I'd consult with @ignopeverell because he introduced tracking of state changes.

Reg Q3 - There are 2 parts - sending request and receiving the response. The first one is easier - replace `unwrap` with proper handling here https://github.com/mimblewimble/grin/blob/master/servers/src/grin/sync.rs#L278  The second part is a bit trickier - my naive attempt was to rely on timer https://github.com/mimblewimble/grin/pull/1320/files#diff-da8815c2e0832a4f8ff91de9eea26c7aR461  because there is no an immediate response , a peer can decide to ignore us or fail to send a file for some reason.

**Comment by garyyu** [comment]
Thanks!
Understood  Q3 now, is this a little bit premature optimization? I never see these errors before on my tests. But the validation fail happened several times on my tests, that's why this fix come.

Anyway, I wait for your consultation result before proceeding for further fix on Q3 :)

**Comment by hashmap** [comment]
it's hard to see the lack of response, basically your node stuck forever on sync without any errors, only restart helps. It happened to me couple times, probably a peer was restarted while handling my  node's request. But I was running fast sync from scratch dozen times per day.

File: servers/src/common/types.rs [hunk 1]
```diff
@@ -296,13 +296,15 @@ pub enum SyncStatus {
 /// Current sync state. Encapsulates the current SyncStatus.
 pub struct SyncState {
 	current: RwLock<SyncStatus>,
+	sync_error: RwLock<Result<String, String>>,
```
[the remark this answers is not in the document]
Comment (garyyu) [reply servers/src/common/types.rs]: Perhaps better to replace `Result<String, String>` with  `Result<String, Error>`?   I used `Error` at the beginning, forgot what problem lead me to replace `Error` with `String` here.

Rust is new to me:(  still confusing about Result, Error, Option and so on.
Are you sure Option is better here than `Result<String, Error>`?

[the remark this answers is not in the document]
Comment (hashmap) [reply servers/src/common/types.rs]: I think so. `Result` is needed when you have something which is value or error - usually it's a result of a computation/function (hence Result). Option is needed when you may have value or nothing (null/nil in many languages), actually Option may sound confusing, in other languages it's called `Maybe` which sounds quite natural - `maybe error`. So `Option<Error>` would  give us 2 possible value - `None` (absence of error) or `Some(error)`.

**Comment by garyyu** [comment]
since it's indeed existing issue, ok to add these 2 fix. is testing by error cases simulation for the new code.

File: servers/src/common/types.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ignopeverell) [reply servers/src/common/types.rs]: I agree with @hashmap that `Option<Error>` would be more idiomatic here.

File: servers/src/common/adapters.rs [hunk 2]
```diff
@@ -361,7 +361,9 @@ impl p2p::ChainAdapter for NetToChainAdapter {
 			w(&self.chain).txhashset_write(h, txhashset_data, self.sync_state.as_ref())
 		{
 			error!(LOGGER, "Failed to save txhashset archive: {}", e);
-			!e.is_bad_data()
+			let is_good_data = !e.is_bad_data();
+			self.sync_state.set_sync_error(types::Error::Chain(e));
```
Comment (ignopeverell) [inline servers/src/common/adapters.rs]: As a less verbose alternative, you can also clone the error here. They're very small objects so it's not a big deal.

File: servers/src/grin/sync.rs [hunk 3]
```diff
@@ -451,6 +458,11 @@ impl SyncInfo {
 			false
 		}
 	}
+
```
Comment (ignopeverell) [inline servers/src/grin/sync.rs]: Perhaps we could get rid of `fast_sync_due` now. The time seems irrelevant after this fix, so we should be able to only rely on the sync state (including the error), right?

File: servers/src/common/adapters.rs (hunk 2 above)
Comment (garyyu) [reply servers/src/common/adapters.rs]: Tried before, unfortunately `Clone is not implemented for failure::Context<error::ErrorKind>`. Have no idea about how to fix it.

And a long story about Error clone,  many Error can't clone, for some reasons such as io::Error, then all Error which packing io::Error can't clone.

File: servers/src/grin/sync.rs (hunk 3 above)
Comment (garyyu) [reply servers/src/grin/sync.rs]: We keep `fast_sync_due` for the moment, use it for download timeout detection. Perhaps switch to better timeout detection method in the future.

**Comment by garyyu** [comment]
Updated PR according to above review comments: 
- Use Option<Error> for 'sync_error' instead of Result. 

And add 2 more fixes according to @hashmap failure cases:
1. handle sending request failure of `send_txhashset_request`;
2. use a `10-mins timeout` to handle no-response failure, in case a peer ignore our txhashset request, or a peer fail to send a file for some reason.

Note about `Error` type: because of the problem of Error `clone` (such as io::Error, all Error which packing io::Error can't clone directly too).  Instead, I prefer `Arc::clone` since it's cross module and multi-threads access. That's why the final `sync_error` becomes:
```
sync_error: Arc::new(RwLock::new(None)),
```

File: servers/src/grin/sync.rs [hunk 4]
```diff
@@ -120,21 +120,51 @@ pub fn run_sync(
...
 						};
 					}
 
 					if fast_sync_enabled {
-						// run fast sync if applicable, every 5 min
-						if header_head.height == si.highest_height && si.fast_sync_due() {
-							fast_sync(peers.clone(), chain.clone(), &header_head);
-							sync_state.update(SyncStatus::TxHashsetDownload);
+
+						// check sync error
```
Comment (hashmap) [inline servers/src/grin/sync.rs]: Nitpicking - a fragment from this line to 151 could be a bit simpler like:
```
						let maybe_sync_error = sync_state.sync_error();
						if let Some(ref sync_error) = *maybe_sync_error.read().unwrap() {
							error!(
								LOGGER,
								"fast_sync: error = {:?}. restart fast sync", sync_error
							);
							si.fast_sync_reset();
							sync_state.clear_sync_error();
						}
```

File: servers/src/common/types.rs [hunk 5]
```diff
@@ -332,6 +334,24 @@ impl SyncState {
 
 		*status = new_status;
 	}
+
+	/// Communicate sync error
+	pub fn set_sync_error(&self, error: Error){
+		let clone = Arc::clone(&self.sync_error);
```
Comment (hashmap) [inline servers/src/common/types.rs]: No clone is needed:
```
	*(self.sync_error.write().unwrap()) = Some(error);
```

File: servers/src/common/types.rs [hunk 6]
```diff
@@ -332,6 +334,24 @@ impl SyncState {
...
+	}
+
+	/// Get sync error
+	pub fn sync_error(&self) -> Arc<RwLock<Option<Error>>> {
+		Arc::clone(&self.sync_error)
+	}
+
+	/// Clear sync error
+	pub fn clear_sync_error(&self){
+		let clone = Arc::clone(&self.sync_error);
```
Comment (hashmap) [inline servers/src/common/types.rs]: Also no clone here:
```
*(self.sync_error.write().unwrap()) = None;
```

File: servers/src/grin/sync.rs (hunk 4 above)
Comment (garyyu) [reply servers/src/grin/sync.rs]: Haha~   You're making same error like me:)  Actually, yours (also my previous trying) can't build, because of an odd short live error, plus a dead-lock risk.

Comment (hashmap) [reply servers/src/grin/sync.rs]: try it:)

**Comment by garyyu** [comment]
@hashmap Thanks again for your review~   The unnecessary clone in set_sync_error/clear_sync_error has been removed.

File: servers/src/grin/sync.rs (hunk 4 above)
Comment (garyyu) [reply servers/src/grin/sync.rs]: I said that's old mistakes what I made before:) it will cause 2 problems there~

Comment (garyyu) [reply servers/src/grin/sync.rs]:
```
error[E0597]: `maybe_sync_error` does not live long enough
   --> servers/src/grin/sync.rs:139:39
    |
139 |                             if let Some(ref sync_error) = *maybe_sync_error.read().unwrap() {
    |                                                            ^^^^^^^^^^^^^^^^ borrowed value does not live long enough
...
160 |                         }
    |                         - `maybe_sync_error` dropped here while still borrowed
    |
    = note: values in a scope are dropped in the opposite order they are created
```

Comment (garyyu) [reply servers/src/grin/sync.rs]: @hashmap   I can add one line after your code to pass the build:
```
drop(maybe_sync_error);
```

But then it's same as my original code:)
And another problem in your code is the `read lock` and the `write lock`, it's a risk to dead lock, I don't know why compiler don't complain about that, but better to avoid.

**Comment by hashmap** [comment]
Thanks for taking care of it!
